In [21]:
import sys

write a function to read in the input text file

In [40]:
def read_input(input_filename):
    '''read the input text file, split the lines, remove the empty last line, and return as string'''
    temp_input = open(input_filename)
    return ''.join(temp_input.read().split('\n')[:-1])

In [41]:
input_2 = read_input('input_p5.txt')
len(input_2)

50000

In [38]:
def destroy_match(string):
    '''recusive method for getting rid of capital letter adjacenting lowercase letter. return the string after the adjacent pairs
    of capital letter and its lowercase are removed'''
    i = 0
    while i < len(string) -1:
        #print('enters while loop')
        if (string[i].lower() == string[i+1] and string[i] != string[i+1]) or (string[i].upper() == string[i+1] and string[i] != string[i+1]):
            #print('enters recursion')
            i += 1
            return destroy_match(string[:i-1] +string[i+1:])
        else:
            #print('enters else clause')
            i += 1
    #print('done with while loop')
    #print(string)
    return string

In [39]:
#test case
destroy_match('AbBacedDEff')

'cff'

In [42]:
#increase the recusion depth limit to handle the big input size
sys.setrecursionlimit(10000)

In [33]:
len(destroy_match(destroy_match(input_2[:10000]) + destroy_match(input_2[10000:20000]) + destroy_match(input_2[20000:30000]) + destroy_match(input_2[30000:40000]) + destroy_match(input_2[40000:])))

10708

Answer to the first problem is 10708 (that is the length of remaining string after the adjacent pair of capital letter and its lowercase are removed).

In [34]:
def remove_char(string,char):
    '''selectively remove alphabetical letter (denote by char) both capital letter and lowercase. and return the resulting string'''
    string = string.replace(char.lower(),'')
    string = string.replace(char.upper(),'')
    return string

In [35]:
output_match = destroy_match(destroy_match(input_2[:10000]) + destroy_match(input_2[10000:20000]) + destroy_match(input_2[20000:30000]) + destroy_match(input_2[30000:40000]) + destroy_match(input_2[40000:]))
len(output_match)

10708

In [43]:
def match_after_remove_letter(string):
    '''keep track of what alphabetical letter is removed, call the pair removal function to assess the remaining letters. keep that information
    and return the alphabetical letter, in which its removal lead to string  that after pair removal is the shortest'''
    
    letters = 'abcdefghijklmnopqrstuvwxyz'
    temp_tuple = ('',len(string))
    for i in letters:
        temp_length = len(destroy_match(remove_char(string,i)))
        if temp_length < temp_tuple[1]:
            temp_tuple =  i,temp_length
    return temp_tuple

In [37]:
match_after_remove_letter(output_match)

('j', 5330)

The answer for 2nd part is 'j'. after remove j, the string can be collapse much further to 5330 remaining characters.